Add aws-k8s-1.23, vmware-k8s-1.23, aws-k8s-1.23-nvidia, metal-k8s-1.23 variants

etungsten opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
Resolves https://github.com/bottlerocket-os/bottlerocket/issues/2139


**Description of changes:**

For reference, the addition of k8s-1.22 variants was in https://github.com/bottlerocket-os/bottlerocket/pull/1962
```
    k8s-1.23: enable `CSIMigration*` feature gates in kubelet
    
    This enables the CSIMigration features gates that enables migration from
    in-tree CSI plugins to solution-specific CSI drivers.
    
    EKS 1.23 is planning to enable CSIMigrationAWS by default.
    
    vSphere container storage plugin migration has been in beta since 1.19
    and the in-tree plugin does not support the newest features in the
    vSphere CSI driver.
    
    Upstream kubernetes is eventually going to deprecate in-tree CSI
    plugins.
```
```
    metal-k8s: limit supported arch to x86_64
```
```
    Add metal-k8s-1.23 variant
    
    This change adds an additional variant `metal-k8s-1.23`, which includes
    necessary Kubernetes packages and settings for running Bottlerocket on
    metal in a Kubernetes v1.23 cluster.
```
```

    Add vmware-k8s-1.23 variant
    
    Adds vmware-k8s-1.23 variant, relinks symlinks in models
    vmware-k8s-1.23 supports boot config settings in models.

    Rename symlink for oci-hooks in vmware-k8s-1.22

```
```
    Add aws-k8s-1.23-nvidia variant
    
    Adds aws-k8s-1.23-nvidia variant, create symlinks for previous k8s
    version nvidia variant in models.
    
    aws-k8s-1.23-nvidia supports boot config settings while other nvidia
    variants do not.
```
```
    Add aws-k8s-1.23 variant
    
    Adds aws-k8s-1.23 variant, relinks symlinks in models.
    aws-k8s-1.23 supports settings.boot while older aws-k8s-* variants do not.
```
```
    containerd,k8s: add new containerd-config templates, update defaults
    
    For K8s 1.23, we're having kubelet use `/run/containerd/containerd.sock`
    for the container runtime socket instead of dockershim.sock.
    
    For compatiblity purposes, K8s 1.23 is maintaining a symlink at
    /run/dockershim.sock. For existing older K8s variants, we're keeping
    dockershim.sock as the main socket for now to avoid breakage until we
    can give sufficient notice of us deprecating dockershim.

```
```
    packages/kubernetes: move ExecStartPre for pause ctr pull to drop-in
    
    This creates a drop-in file for pulling the pause container image with
    host-ctr in `kubelet.service`.
```
```
    packages: add kubernetes-1.23
    
    This adds K8s 1.23 package. We're using K8s 1.23 source from EKS-D.
    
    1.23 Kubelet uses `/run/containerd/containerd.sock` as the container
    runtime socket due to the upcoming dockershim.sock deprecation.
    We also create a symlink for the container runtime socket to be
    `/run/dockershim.sock` for compatibility purposes.
```


**Testing done:**

Testing the containerd socket changes. In 1.23, kubelet starts up fine:
```
bash-5.1# systemctl status kubelet
● kubelet.service - Kubelet
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/kubelet.service; enabled; vendor preset: enabled)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/kubelet.service.d
             └─dockershim-symlink.conf
             /etc/systemd/system/kubelet.service.d
             └─exec-start.conf
     Active: active (running) since Fri 2022-06-03 20:33:17 UTC; 6min ago
       Docs: https://github.com/kubernetes/kubernetes
    Process: 1640 ExecStartPre=/sbin/iptables -P FORWARD ACCEPT (code=exited, status=0/SUCCESS)
    Process: 1645 ExecStartPre=/usr/bin/host-ctr --containerd-socket=/run/containerd/containerd.sock --namespace=k8s.io pull-image --source=${POD_INFRA_CONTAINER_IMAGE} --registry-config=/etc/host-containers/host-ctr.toml (code=exited, status=0/SUCCESS)
    Process: 1669 ExecStartPre=/bin/ln -sf /run/containerd/containerd.sock /run/dockershim.sock (code=exited, status=0/SUCCESS)
   Main PID: 1670 (kubelet)
      Tasks: 16 (limit: 9168)
     Memory: 190.1M
        CPU: 3.420s
     CGroup: /runtime.slice/kubelet.service
             └─1670 /usr/bin/kubelet --cloud-provider aws --kubeconfig /etc/kubernetes/kubelet/kubeconfig --config /etc/kubernetes/kubelet/config --container-runtime=remote --container-runtime-endpoint=unix:///run/containerd/containerd.sock --containerd=/run/containerd/containerd.sock --network-plugin cni --root-dir /var/lib/kubelet --cert-dir /var/lib/kubelet/pki --node-ip 2600:1f14:fb6:1202:8cc4:826:20a6:1334 --node-labels "" --register-with-taints "" --pod-infra-container-image 602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/pause:3.1-eksbuild.1
```

`dockershim.sock` a symlink to `containerd.sock` as expected and is functional:
```
bash-5.1# ls -al /run/dockershim.sock 
lrwxrwxrwx. 1 root root 31 Jun  3 20:33 /run/dockershim.sock -> /run/containerd/containerd.sock
bash-5.1# ctr -a /run/dockershim.sock -n k8s.io containers ls
CONTAINER                                                           IMAGE                                                                                  RUNTIME                  
22f11405fe002530e79d4dc496f1f4bde565f4ab50343d81a9dc5143cc922880    602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-proxy:v1.22.6-eksbuild.1         io.containerd.runc.v2    
...
bash-5.1# ctr -a /run/containerd/containerd.sock -n k8s.io containers ls
CONTAINER                                                           IMAGE                                                                                  RUNTIME                  
22f11405fe002530e79d4dc496f1f4bde565f4ab50343d81a9dc5143cc922880    602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-proxy:v1.22.6-eksbuild.1         io.containerd.runc.v2    
...
```


- [x] Pods run OK
- [x]  Conformance testing with 1.23 control plane and x86_64 aws-k8s-1.23 nodes
```
      "plugin": "e2e",
      "node": "global",
      "status": "complete",
      "result-status": "passed",
      "result-counts": {
        "passed": 346,
        "skipped": 6698
      },
```
- [x]  Conformance testing with 1.23 control plane and aarch64 aws-k8s-1.23 nodes
```
      "plugin": "e2e",
      "node": "global",
      "status": "complete",
      "result-status": "passed",
      "result-counts": {
        "passed": 346,
        "skipped": 6698
      },
```
- [x]  Conformance testing with 1.23 control plane and x86_64 aws-k8s-1.23 nodes in an IPv6 cluster
```
      "plugin": "e2e",
      "node": "global",
      "status": "complete",
      "result-status": "passed",
      "result-counts": {
        "passed": 346,
        "skipped": 6698
      },
```
- [x]  Conformance testing with 1.23 control plane and x86_64 vmware-k8s-1.23 nodes
```
Sonobuoy has completed. Use `sonobuoy retrieve` to get results.
Plugin: e2e
Status: passed
Total: 7044
Passed: 346
Failed: 0
Skipped: 6698

```
- [x] Confirm the EBS CSI driver still works.
```
$ kubectl get pvc block-claim 
NAME          STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
block-claim   Bound    pvc-6765b493-84b3-427e-9bd9-45b9dae0444b   4Gi        RWO            ebs-sc         5s
```
- [x] Confirm the vSphere CSI driver still works
- [x] Smoke test aws-k8s-1.23-nvidia variant with some GPU workloads
```
$ kubectl get jobs -o wide
NAME                COMPLETIONS   DURATION   AGE     CONTAINERS   IMAGES                                                          SELECTOR
nvidia-smoke-test   1/1           113s       2m42s   gpu-tests    359404537045.dkr.ecr.us-west-2.amazonaws.com/gpu-tests:latest
```
- [x] metal K8s 1.23 testing with 1.22 control plane
```
$ ./baremetal-conformance-test.sh   --control-plane-ip    --kubeconfig ./cluster-kubeconfigs/br-test-122-eks-a-cluster.kubeconfig   --hardware-manifest br-eksa-122-nodes.csv   --os-image-url  bottlerocket-metal-k8s-1.23-x86_64.img.gz  --test-mode certified-conformance
Generating hardware manifests for the selected machines...
Getting necessary information for Bottlerocket nodes to join the test cluster...
Creating K8s bootstrap token for node registration...
Creating tinkerbell workflow template for provisioning Bottlerocket...
Waiting for tinkerbell workflows to finish...
Resetting eksa-node10's power...
Resetting eksa-node15's power...
Resetting eksa-node17's power...
Waiting for launched Bottlerocket nodes to become 'Ready' in br-test-122 cluster...
eksa-node10 eksa-node15 eksa-node17 are ready in br-test-122.
Starting Sonobuoy Kubernetes conformance test! Full 'certified-conformance' tests may take up to 1.5 hours to finish
...
   PLUGIN     STATUS   RESULT   COUNT
      e2e   complete   passed       1

Sonobuoy has completed. Use `sonobuoy retrieve` to get results.
Plugin: e2e
Status: passed
Total: 6434
Passed: 346
Failed: 0
Skipped: 6088
Sonobuoy test results available at br-test-122-conformance-test-results/br-test-122-conformance-results.tar.gz
```
- [x] Check kubelet logs and system logs in general, compared them with the aws-k8s-1.23 variant kubelet logs:


... Others TBD


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

etungsten committed:
packages: add kubernetes-1.23

This adds K8s 1.23 package. We're using K8s 1.23 source from EKS-D.

1.23 Kubelet uses `/run/containerd/containerd.sock` as the container
runtime socket due to the upcoming dockershim.sock deprecation.
We also create a symlink for the container runtime socket to be
`/run/dockershim.sock` for compatibility purposes.

etungsten committed:
Add aws-k8s-1.23 variant

Adds aws-k8s-1.23 variant, relinks symlinks in models.
aws-k8s-1.23 supports settings.boot while older aws-k8s-* variants do not.

etungsten committed:
Add aws-k8s-1.23-nvidia variant

Adds aws-k8s-1.23-nvidia variant, create symlinks for previous k8s
version nvidia variant in models.

aws-k8s-1.23-nvidia supports boot config settings while other nvidia
variants do not.

etungsten committed:
Add vmware-k8s-1.23 variant

Adds vmware-k8s-1.23 variant, relinks symlinks in models
vmware-k8s-1.23 supports boot config settings in models.

Rename symlink for oci-hooks in vmware-k8s-1.22

etungsten committed:
Add metal-k8s-1.23 variant

This change adds an additional variant `metal-k8s-1.23`, which includes
necessary Kubernetes packages and settings for running Bottlerocket on
metal in a Kubernetes v1.23 cluster.

etungsten commented:
This is currently blocked  by https://github.com/bottlerocket-os/bottlerocket-sdk/pull/72. We need go1.17 in the `bottlerocket-sdk` in order to build kubelet-1.23.

gthao313 commented:
Just reminder that we have a new kubelet settings [pod-pids-limit](https://github.com/bottlerocket-os/bottlerocket/commit/fbc65856ab0b9d333eb59db415952c6787d2cf75), and you probably need add it to the kube-config file as well : )

zmrow commented on variants/metal-k8s-1.23/Cargo.toml:
```diff
@@ -0,0 +1,41 @@
+partition-plan = "unified"
+
+[package.metadata.build-variant]
+image-format = "raw"
```
Let's add `supported-arches = ["x86_64"]` here

bcressey commented on variants/metal-k8s-1.23/Cargo.toml:
```diff
@@ -0,0 +1,41 @@
+    "kubelet-1.23",
+    "release",
+]
+grub-features = [
+    "set-private-var"
+]
```
I would also like the `*-k8s-1.23` variants to support `settings.boot` since we're certain to have the GRUB changes we need.

etungsten commented:
Push above rebases onto develop to pull in changes for the new SDK

etungsten committed:
metal-k8s: limit supported arch to x86_64

etungsten commented:
Push above adds boot config settings support for all new k8s-1.23 variants.
Push above also limits the supported-architecture for all metal variants to x86_64.

etungsten commented:
Push above bumps the k8s-1.23 to the latest EKS-D dev release.
Push above also temporarily sets `fetch-upstream` to true in the github actions build workflow for k8s-1.23 variants.

etungsten committed:
k8s-1.23: enable `CSIMigration*` feature gates in kubelet

This enables the CSIMigration features gates that enables migration from
in-tree CSI plugins to solution-specific CSI drivers.

EKS 1.23 is planning to enable CSIMigrationAWS by default.

vSphere container storage plugin migration has been in beta since 1.19
and the in-tree plugin does not support the newest features in the
vSphere CSI driver.

Upstream kubernetes is eventually going to deprecate in-tree CSI
plugins.

etungsten commented:
Push above adds missing netdog kernel parameters for configuring network interfaces for AWS/VMware variants.
Push above adds `podPidsLimit` configuration to k8s-1.23 kubelet config.

etungsten commented:
Push above rebases onto develop

etungsten committed:
containerd,k8s: add new containerd-config templates, update defaults

For K8s 1.23, we're having kubelet use `/run/containerd/containerd.sock`
for the container runtime socket instead of dockershim.sock.

For compatiblity purposes, K8s 1.23 is maintaining a symlink at
/run/dockershim.sock. For existing older K8s variants, we're keeping
dockershim.sock as the main socket for now to avoid breakage until we
can give sufficient notice of us deprecating dockershim.

etungsten commented:
Push above adds a new commit: "containerd,k8s: add new containerd-config templates, update defaults"

All new K8s-1.23 variants shall use `/run/containerd/containerd.sock` as the container runtime socket and maintain `/run/dockershim.sock` as a symlink for compatibility purposes.

etungsten commented:
Push above changes `fetch-upstream` to false in github actions now that I've uploaded the k8s 1.23 source tarball to our lookaside cache.

etungsten commented:
Push above rebases onto develop to pull in the latest containerd update and changes.

etungsten commented:
Push above rebases onto develop to pull in changes from https://github.com/bottlerocket-os/bottlerocket/pull/2176
Push above updates the kubelet-config for kubernetes-1.23 to include the latest changes concerning `cluster-dns-ip` rendering.

cbgbt reviewed (approved):
This LGTM, so long as we've tested `settings.boot`. Nice work.

etungsten commented:
> This LGTM, so long as we've tested settings.boot. Nice work.

Very nice catch. I had forgotten to include the defaults for settings.boot in the new variants. Push above fixes that.

arnaldo2792 commented on sources/models/src/aws-k8s-1.23-nvidia/defaults.d/50-kubernetes-aws.toml:
```diff
@@ -0,0 +1 @@
+../../../shared-defaults/kubernetes-aws.toml
```
For the lack of a better place to add this, you are missing a configuration file introduced by #2076:

```
containerd-cri-pki.toml
```

bcressey replied:
My comments were mostly nits - this is the only blocker I see.

bcressey commented on sources/logdog/conf/logdog.aws-k8s-1.23.conf:
```diff
@@ -0,0 +1 @@
+aws-k8s.conf
```
Not for this PR, but we could potentially use our new `bottlerocket-variant` crate to set up a ladder of different configs, ordered from most to least specific, and pick the first one that existed.

That would let us avoid at least this one symlink chore.

bcressey commented on sources/models/src/vmware-k8s-1.23/defaults.d/70-oci-hooks.toml:
```diff
@@ -0,0 +1 @@
+../../../shared-defaults/oci-hooks.toml
```
nit: I liked the non-colliding numbers change made for `metal`:
```
70-public-ntp.toml
80-oci-hooks.toml
90-boot.toml
```

vs.
```
70-oci-hooks.toml
70-public-ntp.toml
80-boot.toml
```

bcressey commented on variants/vmware-k8s-1.23/template.ovf:
```diff
@@ -0,0 +1,79 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<Envelope xmlns="http://schemas.dmtf.org/ovf/envelope/1" xmlns:ovf="http://schemas.dmtf.org/ovf/envelope/1" xmlns:vmw="http://www.vmware.com/schema/ovf" xmlns:rasd="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData" xmlns:vssd="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_VirtualSystemSettingData">
```
This can just be a symlink to `shared/template-unified.ovf`, like `vmware-k8s-1.22`.

bcressey commented on packages/kubernetes-1.23/dockershim-symlink.conf:
```diff
@@ -0,0 +1,2 @@
+[Service]
+ExecStartPre=/bin/ln -sf /run/containerd/containerd.sock /run/dockershim.sock
```
Nice, this is a pretty elegant way to do it. I like that it minimizes the delta between `kubelet.service` across variants.

Not a blocker, but you could potentially do the same thing with the `host-ctr` pull command. If you moved that to a drop-in then the hashes for `kubelet.service` would match across packages.

(I use `sha512sum packages/kubernetes-*/<file>` as a quick way to check for differences when reviewing these PRs. And yes, we should probably have a way to share these files between packages.)

bcressey reviewed (commented):
🎉

zmrow reviewed (approved):
Once @bcressey 's suggestions are addressed I think this looks good!

:safety_vest:

etungsten commented:
Push above rebases onto develop and pulls in changes from https://github.com/bottlerocket-os/bottlerocket/pull/2076 and applies them to 1.23 variants.

etungsten committed:
packages/kubernetes: move ExecStartPre for pause ctr pull to drop-in

This creates a drop-in file for pulling the pause container image with
host-ctr in `kubelet.service`.

etungsten commented:
Push above fixes the missing `[Service]` section in the drop-in file for prepulling pause ctr.

etungsten commented:
I missed that `metal-k8s-1.22` and `metal-k8s-1.21` still needs to use dockershim as the underlying container runtime socket. The push above fixes that.

bcressey commented on packages/kubernetes-1.21/prestart-pull-pause-ctr.conf:
```diff
@@ -0,0 +1,8 @@
+    --namespace=k8s.io \
+    pull-image \
+    --source=${POD_INFRA_CONTAINER_IMAGE} \
+    --registry-config=/etc/host-containers/host-ctr.toml
```
nit: GitHub is showing a different line ending here, vs. 1.19 or 1.20, which throws off my approach of using hashes to check for differences.

bcressey reviewed (approved):
🥇

etungsten commented:
Push above adds missing newline at the end of `packages/kubernetes-1.21/prestart-pull-pause-ctr.conf`

vasireddy99 commented:
Hi @etungsten, Can you please confirm if this change/support is included in the latest release `v1.8`. if so can you please update it in the `release notes`.

etungsten commented:
@vasireddy99, K8s 1.23 variants are included in v1.8.0 and it's also listed in the release notes here under the "Kubernetes" section: https://github.com/bottlerocket-os/bottlerocket/releases/tag/v1.8.0

vasireddy99 commented:
Sure @etungsten, Yeah I can see that and so helpful 🙂.... But I think this PR #2140 would be appropriate as well to be there. since, there is a change in the runtime to containerd which was not the case in previous supported versions such as for  kubernetes 1.22.
